### PR TITLE
[SPARK-49269][SQL] Eagerly evaluate VALUES() list in AstBuilder

### PR DESCRIPTION
**Pull request by costas-db**
### What changes were proposed in this pull request?
This is a continuation of a prior performance improvement: https://github.com/apache/spark/pull/47428 that eagerly evaluates memory-heavy `UnresolvedUnlineTables` parse tree nodes as soon as they are constructed in the AstBuilder. 

This PR applies this optimization to any statement that might contain one or more `VALUES()` clauses (such as subqueries etc), instead of just applying that optimization to `INSERT INTO ... VALUES` statements, which is what the prior PR did.

### Why are the changes needed?
With these changes we not only reduce the memory footprint of every statement that can contain the `VALUES()` clause, but we also improve upon the previous optimization as we avoid unnecessary traversals of the parse tree, which not only improves the runtime performance, but also minimizes the amount of time in which the `UnresolvedInlineTable` is kept in memory. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Provided scala tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by dtenedor — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
Ill merge after CI passes

File: sql/core/src/test/scala/org/apache/spark/sql/InlineTableParsingImprovementsSuite.scala [hunk 1]
```diff
@@ -226,4 +287,72 @@ class InlineTableParsingImprovementsSuite extends QueryTest with SharedSparkSess
       }
     }
   }
+
+  test("Value list in subquery") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/InlineTableParsingImprovementsSuite.scala:291]: Let's add a JIRA prefix in the test title

```suggestion
  test("SPARK-49269: Value list in subquery") {
```

File: sql/core/src/test/scala/org/apache/spark/sql/InlineTableParsingImprovementsSuite.scala [hunk 2]
```diff
@@ -226,4 +287,72 @@ class InlineTableParsingImprovementsSuite extends QueryTest with SharedSparkSess
...
+        if (flagVals.head == eagerEvalOfUnresolvedInlineTableEnabled) {
+          firstDF = Some(res)
+        } else {
+          checkAnswer(res, firstDF.get)
+        }
+      }
+    }
+  }
+
+  test("Value list in projection list subquery") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/InlineTableParsingImprovementsSuite.scala:325]: here too

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**